academic-services/knowledge-exploration-service/tutorial-grammar-design.md

```diff
@@ -0,0 +1,392 @@
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_create_return_query":::
+
+### Support multiple attribute search
+
+Multiple attribute search is used for processing exploratory search queries that contain multiple attribute terms. For example, a user may want to search for publications given a few fields of study and affiliations.
+
+We use the repeat item `<item repeat="1-INF" repeat-logprob="-1">` element to process queries that contains multiple attributes by continously matching remaining query terms using different search grammars.
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_multiple_attribute_search_loop":::
+
+This allows us to process queries such as "papers from microsoft about machine learning"
+
+### Support partial attribute search
+
+Partial attribute search is used for processing keyword based search queries. For example, a user may search for a paper using terms from the paper's title or abstract.
+
+We use a `<item repeat="1-INF">` element to process keyword based queries using abstract and title words.
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_partial_attribute_search":::
+
+Notice that we have heavier penalties (-3 and -4) associated with title and abstract word search compared to other grammar rules in the sample grammar. (i.e. author and affiliation searches have a penalty of -2). This is designed to demote title/abstract keyword searches if more specific attribute searches yield results.
+
+In addition to heavier penalties, we also set a minimum word match count requirement for title and abstract words search. Title and abstract words may cover lots of common terms that can be matched against terms in a query. We introduce this requirement to create a higher quality bar for title and abstract words search results. The following code ensures that the title and abstract words search is only valid if there are 3 or more words in the query that can be matched against the title/abstract words.
+
+:::code language="xml" source="samplePrivateLibraryData.linked.search.grammar.xml" id="snippet_partial_attribute_search_constraints":::
+
+### Support drop terms with penalties
+
+We can improve multiple attribute and partial attribute search by allowing drop/garbage terms in the query. For example, a user may want to search for the publication "A Review of Microsoft Academic Services for Science of Science Studies" by using the parts of the title the user recalls and enters the query "Microsoft Academic Services". Having drop terms support allows this type of queries to be searchable by partial title search.
```

Comment:
>Having drop terms support allows this type of queries to be searchable by partial title search. [](start = 341, length = 95)

"Allowing query terms to be ignored enables this type of query to be matched via keyword attributes."